packages/context/src/context.ts

```diff
@@ -64,65 +112,170 @@ export class Context {
     return this.registry.has(key);
   }
 
+  /**
+   * Iterate all contexts by breadth first traversal of the graph following
+   * `parents`. For example, with the following context graph:
+   * - reqCtx -> [serverCtx, connectorCtx]
+   * - serverCtx -> [appCtx]
+   * - connectorCtx -> [appCtx]
+   * `reqCtx.getAllContexts()` returns an iterator of `[reqCtx, serverCtx,
+   * connectorCtx, appCtx]`.
+   *
+   * @param visitor A function to produce a result from the context object
+   * locally without consulting the parents.
+   */
+  protected *contextIterator<T>(
+    visitor: (ctx: Context) => {done: boolean; value: T},
```

Comment:
Methods should start with a verb - e.g. `createContextIterator` or `iterateContextHierarchy`.

I have mixed feelings about the proposed design of this method, because it combines two things into a single method:
 - iterate contexts
 - map contexts to values

In my experience, this often leads to situations where we want one feature (e.g. iterate contexts) but not the other (map contexts via a visitor function), and therefore have troubles using this 2-in-1 function.

I am proposing to decouple the iteration from map (and reduce function below):

Change this function to return an object conforming to standard ES6 [Iteration protocol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols). This allows consumers of this API to use a wider variety of functions for processing the list of contexts.

```ts
protected listContexts() {
  return {
    *[Symbol.iterator]() {
      // the body of your "contextIterator" function goes here
    }
  };
}
```

Users wishing to map contexts to a different value should use `for of`, `Array.from([...contexts]).map()` or a functional-programming-style alternative like [map-interable](https://github.com/parro-it/map-iterable/blob/511d8c00383f851c5f1be242767649a0072db84d/index.js).

```ts
// option 1
const names = [];
for (c of ctx.listContexts()) names.push(c.name);

// option 2
const names = Array.from(ctx.listContexts()).map(c => c.name);

// option 3 (no temporary array is created)
const names = mapIterable(c => c.name, ctx.listContexts());
```

Related reading: http://raganwald.com/2015/02/17/lazy-iteratables-in-javascript.html#iterables-out-to-infinity

---

To be honest, considering that we expect the context hierarchy to be small (let's say ~10 contexts at most), I think returning an Iterable object is over-engineering here (a preliminary optimization?) and this method should simply construct and return an array instance.

I think something like the following may work?

```ts
protected listContexts() {
  const sorted: Set<Context> = new Set();
  const queue: Context[] = [];
  const result: Context[] = [];
  queue.push(this);
  while (queue.length) {
    const c = queue.shift()!;
    if (sorted.has(c)) continue;
    result.push(c);
    queue.push(...c.parents);
  }
  return result;
}
```

Reply:
The generator function returns `IteratableIterator` and it's ready for `for of` out of box.

Reply:
I had an array based implementation at first, but was allured to try the generator function to create an iterator especially as some visitors don't need to go over all contexts.

In our case, I think either way should be fine.